src/ensemble_scheduler/ensemble_scheduler.cc

```diff
@@ -603,29 +642,57 @@ void
 EnsembleContext::RequestComplete(
     TRITONSERVER_InferenceRequest* request, const uint32_t flags, void* userp)
 {
-  if ((flags & TRITONSERVER_REQUEST_RELEASE_ALL) != 0) {
-    LOG_TRITONSERVER_ERROR(
-        TRITONSERVER_InferenceRequestDelete(request),
-        "deleting ensemble inference request");
-    auto request_tracker = reinterpret_cast<RequestTracker*>(userp);
-    if (request_tracker->DecrementCounter()) {
-      delete request_tracker;
+  auto request_tracker = reinterpret_cast<RequestTracker*>(userp);
+  auto pool = request_tracker->CallbackPool();
+  auto fn = [request, flags, request_tracker]() {
+    if ((flags & TRITONSERVER_REQUEST_RELEASE_ALL) != 0) {
+      LOG_TRITONSERVER_ERROR(
+          TRITONSERVER_InferenceRequestDelete(request),
+          "deleting ensemble inference request");
+      if (request_tracker->DecrementCounter()) {
+        delete request_tracker;
+      }
     }
+  };
+
+  // Attempt to enqueue the callback. If all workers are busy and queue is at
+  // capacity, execute the callback immediately in current thread.
+  if (pool->TaskQueueSize() < pool->Size()) {
```

Comment:
Is this correct? `Size()` returns the number of worker and `TaskQueueSize()` returns the number of "pending" task. You can consider the workers are busy when `TaskQueueSize() > 0`, because `pool->TaskQueueSize() == pool->Size()` actually means the # of pending requests equals to the # of workers, right?